### PR TITLE
ISSUES-1885 UTF8 countCodePoints use SIMD

### PR DESCRIPTION
**Pull request by zhang2014**
#1885
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Something is wrong: https://travis-ci.org/yandex/ClickHouse/jobs/339794280#L6819

File: dbms/src/Common/UTF8Helpers.h [hunk 1]
```diff
@@ -49,9 +52,37 @@ inline size_t seqLength(const UInt8 first_octet)
...
+
+    const auto upper_bound = _mm_set1_epi8(0x7F + 1);
+    const auto lower_bound = _mm_set1_epi8(0xC0 - 1);
+
+    for (; data < src_end_sse;)
+    {
+        UInt8 mem_res[16] = {0};
+        auto sse_res = _mm_set1_epi8(0);
+
+        for (int i = 0; i < 0XFF && data < src_end_sse; ++i, data += bytes_sse)
```
Comment (alexey-milovidov) [inline dbms/src/Common/UTF8Helpers.h]: Why we have capital X here? That's correct but looks strange.

File: dbms/src/Common/UTF8Helpers.h [hunk 2]
```diff
@@ -49,9 +52,37 @@ inline size_t seqLength(const UInt8 first_octet)
...
+        {
+            const auto chars = _mm_loadu_si128(reinterpret_cast<const __m128i *>(data));
+            sse_res = _mm_add_epi8(sse_res,
+                                   _mm_or_si128(_mm_cmplt_epi8(chars, upper_bound),
```
Comment (alexey-milovidov) [inline dbms/src/Common/UTF8Helpers.h]: It's interesting if we can do subtract (add) with overflow (to align a range with zero) and then do just single comparison.

File: dbms/src/Common/UTF8Helpers.h (hunk 1 above)
Comment (alexey-milovidov) [inline dbms/src/Common/UTF8Helpers.h]: What is `i`. And why it is limited?

File: dbms/src/Common/UTF8Helpers.h [hunk 3]
```diff
@@ -49,9 +52,37 @@ inline size_t seqLength(const UInt8 first_octet)
...
+                                                _mm_cmpgt_epi8(chars, lower_bound)));
+        }
+
+        _mm_store_si128(reinterpret_cast<__m128i *>(mem_res), sse_res);
+
+        for (auto count : mem_res)
```
Comment (alexey-milovidov) [inline dbms/src/Common/UTF8Helpers.h]: That's fine.

But if you want to do some hard way,
you can search for suitable horizontal add intrinsics in
https://software.intel.com/sites/landingpage/IntrinsicsGuide/

**Comment by zhang2014** [comment]
OK, I may need some time,because I can't use the computer tomorrow

File: dbms/src/Common/UTF8Helpers.h (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Common/UTF8Helpers.h]: Ok, I understand.

**Comment by alexey-milovidov** [comment]
You should to add a case in performance tests.
Look at `tests/performance`.

File: dbms/src/Common/UTF8Helpers.h [hunk 4]
```diff
@@ -49,9 +56,46 @@ inline size_t seqLength(const UInt8 first_octet)
...
+
+            ///Align to zero for the solve two case
+            const auto align_res = _mm_add_epi8(chars, align_sse);
+            ///Because _mm_cmmpgt_epid8 return 0xFF if true
+            const auto choose_res = _mm_blendv_epi8(zero_sse, one_sse, _mm_cmpgt_epi8(align_res, upper_bound));
+
+            sse_res = _mm_add_epi8(sse_res, choose_res);
+        }
+
+        UInt8 mem_res[16] = {0};
```
Comment (zhang2014) [inline dbms/src/Common/UTF8Helpers.h]: @alexey-milovidov Sorry, I flushed your comments. 
Yes, I'm not sure it's is better. Do you have any Suggestions for me?

File: dbms/src/Common/UTF8Helpers.h [hunk 5]
```diff
@@ -49,9 +56,46 @@ inline size_t seqLength(const UInt8 first_octet)
...
+
+            ///Align to zero for the solve two case
+            const auto align_res = _mm_add_epi8(chars, align_sse);
+            ///Because _mm_cmmpgt_epid8 return 0xFF if true
+            const auto choose_res = _mm_blendv_epi8(zero_sse, one_sse, _mm_cmpgt_epi8(align_res, upper_bound));
+
+            sse_res = _mm_add_epi8(sse_res, choose_res);
+        }
+
+        UInt16 mem_res[8] = {0};
```
Comment (zhang2014) [inline dbms/src/Common/UTF8Helpers.h]: @alexey-milovidov  Sorry, I flushed your comments.
Yes, I'm not sure it's better.Can you give me some advice? Thanks

**Comment by zhang2014** [comment]
Is there documentation on how to add a performance test and see it ?

**Comment by alexey-milovidov** [comment]
How to add performance test?

Look at examples in `tests/performance`.
You can add your test case in .xml file, possibly in separate directory.

Performance test is a query or multiple queries, or a query with macro-substitutions,
along with run mode (run infinite query, run query in a loop),
stop conditions (example: max execution speed doesn't change for some amount of time or iterations),
and list of metrics (example: min execution time).

Then you can run it with `clickhouse-performance-test` or `clickhouse performance-test`.
It will send queries to running ClickHouse instance and provide a report.

First you should think about test cases: what queries will probably be affected by your changes?

**Comment by alexey-milovidov** [comment]
In performance test, you should have a representative set of cases for different string sizes and for different ratio of non-ASCII characters. Also it's better to get realistic cases from real dataset... but first we will use only synthetic test cases.

The queries can look like this:

```
WITH s = toString(number) SELECT * FROM system.numbers WHERE NOT ignore(lengthUTF8({s}));
```

where `{s}` is a macro substitution (that will generate many test cases).

`{s}` can be something like:

Short strings:
```
materialize('') 
materialize('Hello, world') 
toString(number)
reinterpretAsString(number)
materialize('中华人民共和国')
```

Medium to long strings:

```
concat('中华人民共和国 ', toString(number), ' Привет, мир!')
concat(concat('中华人民共和国 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)
```

Execution time will be dominated by concat, not lengthUTF8... but performance test should show some difference.

**Comment by alexey-milovidov** [comment]
Running performance test is as simple as

```
clickhouse performance-test .
```

in the directory with your test.

**Comment by zhang2014** [comment]
change before:
``` code javascript
Running: functions_length
[
{
    "hostname": "bogon",
    "main_metric": "avg_rows_per_second",
    "num_cores": 4,
    "num_threads": 8,
    "parameters": {
        "string": ["materialize('')", "materialize('Hello, world')", "toString(number)", "reinterpretAsString(number)", "materialize('中文测试字符串')", "materialize('https://github.com/yandex/ClickHouse/pull/1882')", "materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')", "concat('中文测试字符串 ', toString(number), ' Привет, мир!')", "concat(concat('中文测试字符串 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)", "convertCharset(concat(reinterpretAsString(rand64(1)), reinterpretAsString(rand64(2)), reinterpretAsString(rand64(3)), reinterpretAsString(rand64(4)), reinterpretAsString(rand64(5)), reinterpretAsString(rand64(6)), reinterpretAsString(rand64(7)), reinterpretAsString(rand64(8)), reinterpretAsString(rand64(9)), reinterpretAsString(rand64(10))), 'UTF-16', 'UTF-8')"]
    },
    "ram": 17179869184,
    "runs": [
        {
            "avg_rows_per_second": 9705457.748381,
            "parameters": {
                "string": "materialize('')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('')))"
        },
        {
            "avg_rows_per_second": 8177312.594839,
            "parameters": {
                "string": "materialize('Hello, world')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('Hello, world')))"
        },
        {
            "avg_rows_per_second": 7448756.390556,
            "parameters": {
                "string": "toString(number)"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(toString(number)))"
        },
        {
            "avg_rows_per_second": 8484560.110418,
            "parameters": {
                "string": "reinterpretAsString(number)"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(reinterpretAsString(number)))"
        },
        {
            "avg_rows_per_second": 6649553.395335,
            "parameters": {
                "string": "materialize('中文测试字符串')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('中文测试字符串')))"
        },
        {
            "avg_rows_per_second": 5469924.048361,
            "parameters": {
                "string": "materialize('https://github.com/yandex/ClickHouse/pull/1882')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('https://github.com/yandex/ClickHouse/pull/1882')))"
        },
        {
            "avg_rows_per_second": 3401270.998975,
            "parameters": {
                "string": "materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')))"
        },
        {
            "avg_rows_per_second": 1727267.914353,
            "parameters": {
                "string": "concat('中文测试字符串 ', toString(number), ' Привет, мир!')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(concat('中文测试字符串 ', toString(number), ' Привет, мир!')))"
        },
        {
            "avg_rows_per_second": 412346.953366,
            "parameters": {
                "string": "concat(concat('中文测试字符串 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(concat(concat('中文测试字符串 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)))"
        },
        {
            "avg_rows_per_second": 370100.266678,
            "parameters": {
                "string": "convertCharset(concat(reinterpretAsString(rand64(1)), reinterpretAsString(rand64(2)), reinterpretAsString(rand64(3)), reinterpretAsString(rand64(4)), reinterpretAsString(rand64(5)), reinterpretAsString(rand64(6)), reinterpretAsString(rand64(7)), reinterpretAsString(rand64(8)), reinterpretAsString(rand64(9)), reinterpretAsString(rand64(10))), 'UTF-16', 'UTF-8')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(convertCharset(concat(reinterpretAsString(rand64(1)), reinterpretAsString(rand64(2)), reinterpretAsString(rand64(3)), reinterpretAsString(rand64(4)), reinterpretAsString(rand64(5)), reinterpretAsString(rand64(6)), reinterpretAsString(rand64(7)), reinterpretAsString(rand64(8)), reinterpretAsString(rand64(9)), reinterpretAsString(rand64(10))), 'UTF-16', 'UTF-8')))"
        }
    ],
    "server_version": "1.1.54350",
    "test_name": "functions_length",
    "time": "2018-02-13 15:58:26"
}
]
```
change after:
```code json
[
{
    "hostname": "bogon",
    "main_metric": "avg_rows_per_second",
    "num_cores": 4,
    "num_threads": 8,
    "parameters": {
        "string": ["materialize('')", "materialize('Hello, world')", "toString(number)", "reinterpretAsString(number)", "materialize('中文测试字符串')", "materialize('https://github.com/yandex/ClickHouse/pull/1882')", "materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')", "concat('中文测试字符串 ', toString(number), ' Привет, мир!')", "concat(concat('中文测试字符串 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)", "convertCharset(concat(reinterpretAsString(rand64(1)), reinterpretAsString(rand64(2)), reinterpretAsString(rand64(3)), reinterpretAsString(rand64(4)), reinterpretAsString(rand64(5)), reinterpretAsString(rand64(6)), reinterpretAsString(rand64(7)), reinterpretAsString(rand64(8)), reinterpretAsString(rand64(9)), reinterpretAsString(rand64(10))), 'UTF-16', 'UTF-8')"]
    },
    "ram": 17179869184,
    "runs": [
        {
            "avg_rows_per_second": 9666943.835546,
            "parameters": {
                "string": "materialize('')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('')))"
        },
        {
            "avg_rows_per_second": 8395953.138103,
            "parameters": {
                "string": "materialize('Hello, world')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('Hello, world')))"
        },
        {
            "avg_rows_per_second": 7429674.250748,
            "parameters": {
                "string": "toString(number)"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(toString(number)))"
        },
        {
            "avg_rows_per_second": 8413304.508868,
            "parameters": {
                "string": "reinterpretAsString(number)"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(reinterpretAsString(number)))"
        },
        {
            "avg_rows_per_second": 6717091.001393,
            "parameters": {
                "string": "materialize('中文测试字符串')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('中文测试字符串')))"
        },
        {
            "avg_rows_per_second": 5345943.894829,
            "parameters": {
                "string": "materialize('https://github.com/yandex/ClickHouse/pull/1882')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('https://github.com/yandex/ClickHouse/pull/1882')))"
        },
        {
            "avg_rows_per_second": 3423176.617490,
            "parameters": {
                "string": "materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')))"
        },
        {
            "avg_rows_per_second": 1718039.320371,
            "parameters": {
                "string": "concat('中文测试字符串 ', toString(number), ' Привет, мир!')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(concat('中文测试字符串 ', toString(number), ' Привет, мир!')))"
        },
        {
            "avg_rows_per_second": 431675.967113,
            "parameters": {
                "string": "concat(concat('中文测试字符串 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(concat(concat('中文测试字符串 ', toString(number), ' Привет, мир!') AS x, x, x, x, x, x, x, x, x, x)))"
        },
        {
            "avg_rows_per_second": 396635.677219,
            "parameters": {
                "string": "convertCharset(concat(reinterpretAsString(rand64(1)), reinterpretAsString(rand64(2)), reinterpretAsString(rand64(3)), reinterpretAsString(rand64(4)), reinterpretAsString(rand64(5)), reinterpretAsString(rand64(6)), reinterpretAsString(rand64(7)), reinterpretAsString(rand64(8)), reinterpretAsString(rand64(9)), reinterpretAsString(rand64(10))), 'UTF-16', 'UTF-8')"
            },
            "query": "SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(convertCharset(concat(reinterpretAsString(rand64(1)), reinterpretAsString(rand64(2)), reinterpretAsString(rand64(3)), reinterpretAsString(rand64(4)), reinterpretAsString(rand64(5)), reinterpretAsString(rand64(6)), reinterpretAsString(rand64(7)), reinterpretAsString(rand64(8)), reinterpretAsString(rand64(9)), reinterpretAsString(rand64(10))), 'UTF-16', 'UTF-8')))"
        }
    ],
    "server_version": "1.1.54350",
    "test_name": "functions_length",
    "time": "2018-02-13 16:08:02"
}
]
```

**Comment by zhang2014** [comment]
autotest

**Comment by alexey-milovidov** [comment]
Looks like you compile with debug mode, because I see different numbers:

```
:) SELECT count() FROM system.numbers WHERE NOT ignore(lengthUTF8(materialize('Hello, world')))

SELECT count()
FROM system.numbers 
WHERE NOT ignore(lengthUTF8(materialize('Hello, world')))

Ok.
Query was cancelled.

0 rows in set. Elapsed: 5.613 sec. Processed 165.09 million rows, 1.32 GB (29.41 million rows/s., 235.29 MB/s.)
```

on very old CPU (E5645).

File: dbms/src/Common/UTF8Helpers.h [hunk 6]
```diff
@@ -49,9 +56,47 @@ inline size_t seqLength(const UInt8 first_octet)
...
+        {
+            const auto chars = _mm_loadu_si128(reinterpret_cast<const __m128i *>(data));
+
+            ///Align to zero for the solve two case
+            const auto align_res = _mm_adds_epu8(chars, align_sse);
+            const auto less_than_and_equals = _mm_cmpeq_epi8(_mm_min_epu8(align_res, upper_bound), align_res);
+            ///Because _mm_cmpeq_epi8 return 0xFF if true
+            const auto choose_res = _mm_blendv_epi8(zero_sse, one_sse, less_than_and_equals);
+
+            sse_res = _mm_adds_epu8(sse_res, choose_res);
```
Comment (alexey-milovidov) [inline dbms/src/Common/UTF8Helpers.h]: Probably it is better to use typical approach with movmask and popcnt.
It looks like this:

```
UInt16 bit_mask = _mm_movemask_epi8(byte_mask);
count += __builtin_popcount(bit_mask);
```

but you should test.

Comment (zhang2014) [reply dbms/src/Common/UTF8Helpers.h]: Yes, it's a better choice. I've changed it.

**Comment by KochetovNicolai** [comment]
Improves for long strings. Before:
```SQL
SELECT count()
FROM system.numbers 
WHERE NOT ignore(lengthUTF8(materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')))

↑ Progress: 223.87 million rows, 1.79 GB (10.90 million rows/s., 87.22 MB/s.) Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 20.639 sec. Processed 223.87 million rows, 1.79 GB (10.85 million rows/s., 86.78 MB/s.)
```
After:
```SQL
SELECT count()
FROM system.numbers 
WHERE NOT ignore(lengthUTF8(materialize('https://zh.wikipedia.org/wiki/%E4%B8%AD%E6%97%A5%E9%9F%93%E7%B5%B1%E4%B8%80%E8%A1%A8%E6%84%8F%E6%96%87%E5%AD%97%E6%93%B4%E5%B1%95%E5%8D%80F')))

↘ Progress: 327.61 million rows, 2.62 GB (18.38 million rows/s., 147.01 MB/s.) Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 17.931 sec. Processed 327.61 million rows, 2.62 GB (18.27 million rows/s., 146.17 MB/s.) 
```

**Pull request merged**